torch/csrc/autograd/engine.cpp

```diff
@@ -1107,7 +1107,8 @@ void Engine::evaluate_function(
           next.input_nr, std::move(output), opt_parent_stream, opt_next_stream);
 
       if (is_ready) {
-        auto queue = ready_queue(cpu_ready_queue, input_buffer.device());
+        // NB: The first forward output's device determines the queue
+        auto queue = ready_queue(cpu_ready_queue, next.function->input_metadata(0).device());
```

Comment:
I think we want to preserve the non-CPU device being prioritized.
Could you update this by passing the input metadatas to the InputBuffer struct and make the .device() method there use that instead of the current buffer?

Reply:
Ah good catch!